sdl-core/src/main/scala/io/smartdatalake/workflow/dataobject/expectation/UniqueKeyExpectation.scala

```diff
@@ -0,0 +1,94 @@
+ * In this case the relative standard deviation of approx_count_distinct is set to a high precision.
+ *
+ * @param key Optional list of key columns to evaluate uniqueness
+ * @param expectation Optional SQL comparison operator and literal to define expected value for validation. Default is '= 1'.
+ *                    Together with the result of the aggExpression evaluation on the left side, it forms the condition to validate the expectation.
+ *                    If no expectation is defined, the aggExpression evaluation result is just recorded in metrics.
+ * @param precision Number of digits to keep when calculating fraction. Default is 4.
+ */
+case class UniqueKeyExpectation(
+                                 override val name: String, key: Seq[String] = Seq(),
+                                 override val expectation: Option[String] = Some("= 1"),
+                                 override val precision: Short = 4,
+                                 override val scope: ExpectationScope = Job,
+                                 override val failedSeverity: ExpectationSeverity = ExpectationSeverity.Error )
+  extends Expectation with ExpectationFractionMetricDefaultImpl {
+  assert(scope != ExpectationScope.JobPartition, "scope=JobPartition not supported by UniqueKeyExpectation for now")
+
+  override def roundFunc(v: Double): Double = math.floor(v) // use floor to be more aggressive on detecting unique key violations.
+
+  override val description: Option[String] = Some("fraction of output count-distinct over output count")
+  private val countDistinctName = "countDistinct" + name.capitalize
+  private val countName = if (scope == ExpectationScope.Job) "count" else "countAll"
+  def getPrimaryKeyCols(dataObjectId: DataObjectId)(implicit context: ActionPipelineContext): Seq[String] = {
+    context.instanceRegistry.get[DataObject](dataObjectId) match {
+      case x: TableDataObject =>
+        assert(x.table.primaryKey.nonEmpty, s"($dataObjectId) 'table.primaryKey' must be defined on DataObject if parameter 'UniqueKeyExpectation.columns' is empty")
+        x.table.primaryKey.getOrElse(throw new IllegalStateException(s"($dataObjectId) 'table.primaryKey' is defined but empty..."))
+      case _ => throw ConfigurationException(s"($dataObjectId) If parameter 'columns' is empty, UniqueKeyExpectation must be defined on a DataObject implementing TableDataObject in order to use primary key")
+    }
+  }
+  override def getAggExpressionColumns(dataObjectId: DataObjectId)(implicit functions: DataFrameFunctions, context: ActionPipelineContext): Seq[GenericColumn] = {
+    val colsToCheck = (if (key.isEmpty) getPrimaryKeyCols(dataObjectId) else key).map(functions.col)
+    Seq(
+      if (scope == ExpectationScope.Job && functions.requestSubFeedType() == typeOf[SparkSubFeed]) Some(functions.approxCountDistinct(functions.struct(colsToCheck:_*), Some(0.005)).as(countDistinctName))
```

Comment:
what does magic number 0.005 mean? is that the "high precision"? could be nice to extract it to a variable

Reply:
Should be calculated from precision -> TODO

Reply:
Mapping from precision to rsd is not possible as one measure is absolute and the other is relative.
I added configuration options "approximate" and "approximateRsd" to enable approximate distinct calculation.
Default is exact distinct calculation, using a separate query to evaluate.